### PR TITLE
Create and test a JsonConverter class that makes a file of JSON objects.

### PR DESCRIPTION
**Pull request by cxfloeder**
I made a JsonConverter class that creates a new file with JSON objects written to it. The file is stored in a data folder. I created a test class that checks that JsonConverter can create a file and that the file contents are what I expect it to be.

### CODE REVIEW

File: src/main/java/data/Account.java [hunk 1]
```diff
@@ -15,100 +15,100 @@
...
+  private String currency;
+  private String direction;
+  private String legacyAccountID;
+  private int nextGenAccountID;
+  private String matchingMode;
+  private String aggregationMode;
+
+  public Account(String accountID, String vendorID, String entity, 
+    String currency, String direction, String legacyAccountID, 
+    int nextGenAccountID, String matchingMode, String aggregationMode) {
```
Comment (cadecrandall) [inline src/main/java/data/Account.java:30]: Can you indent these +2 spaces?

Comment (cxfloeder) [reply src/main/java/data/Account.java:30]: It looks like line 29 & 30 already indented 2 spaces.

File: src/main/java/data/JsonConverter.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    File file = null;
+    try {
+      file = new File(fileName);
+      BufferedWriter out = new BufferedWriter(new FileWriter(file));
+      out.write(parentObject.toString());
+      out.close();
+    } catch(IOException e) {
+      return null;
+    } 
+
```
Comment (cadecrandall) [inline src/main/java/data/JsonConverter.java:47]: Did you consider using the GSON library to handle conversions between JSON?

File: src/test/java/javatests/JsonConverterTest.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    try {
+      input = new Scanner(file);
+    } catch(FileNotFoundException e) {
+      input = null;
+    }
+
+    while(input.hasNextLine())
+    {
+      fileContent += input.nextLine();
+    }
```
Comment (cadecrandall) [inline src/test/java/javatests/JsonConverterTest.java:51]: nit: can you put these braces in [K&R style?](https://google.github.io/styleguide/javaguide.html#s4.1.2-blocks-k-r-style)

File: src/test/java/javatests/JsonConverterTest.java [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+    // Test that the file isn't empty and has the right content.
+    assertTrue("createFile() didn't add any content to the file when is should've",
+      !fileContent.equals(""));
```
Comment (cadecrandall) [inline src/test/java/javatests/JsonConverterTest.java:55]: Would it make sense for this test to be assertFalse? I found the logic confusing to understand at first, but not sure if the change makes sense

File: src/main/java/data/JsonConverter.java [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    return file;
+  }
+
+  public boolean updateFile(Vendor vendor) {
+    createFile();
+    return true;
+  }
```
Comment (cadecrandall) [inline src/main/java/data/JsonConverter.java:54]: does this method return false if the file can't be created?

File: src/main/java/data/Account.java (hunk 1 above)
Comment (cadecrandall) [reply src/main/java/data/Account.java:30]: Sorry, I meant two more spaces following the linewrap

File: src/test/java/javatests/JsonConverterTest.java (hunk 3 above)
Comment (cxfloeder) [reply src/test/java/javatests/JsonConverterTest.java:51]: Fixed.

File: src/test/java/javatests/JsonConverterTest.java (hunk 4 above)
Comment (cxfloeder) [reply src/test/java/javatests/JsonConverterTest.java:55]: Yes, that would make sense. I am just used to making every test assertTrue. I will change it to assertFalse.

File: src/main/java/data/JsonConverter.java (hunk 5 above)
Comment (cxfloeder) [reply src/main/java/data/JsonConverter.java:54]: It will once I add more functionality to it. Later, if there is an error creating a file, the method will return false signifying to you that there was a problem with the JsonConverter Component.

File: src/main/java/data/JsonConverter.java (hunk 2 above)
Comment (cxfloeder) [reply src/main/java/data/JsonConverter.java:47]: No, I didn't consider that. My thought process was that I needed a way to create Json objects representing Vendors and Accounts and then take those objects and write them to a new file. Handling the conversion to JSON using GSON may be more difficult than the way I am currently using.

File: src/main/java/data/JsonConverter.java [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+import java.io.BufferedWriter;
+import java.io.IOException;
+import java.util.ArrayList;
+import org.json.JSONObject;
+
+/** A class that creates a config file from a Vendor object. */
+public class JsonConverter {
+  private static final String FILE_PATH_BASE = "data/";
+  
+  public File createFile() {
```
Comment (jakecarlson1) [inline src/main/java/data/JsonConverter.java:28]: The design doc calls out a FileUtil class for writing a file. My impression was that, this JsonConverter class would handle serializing Account/Customer objects to a JSON string. Then, FileUtil would call JsonConverter to get the string representation of an object, and handle the file IO necessary to write this to the file system.

I see that the Json Converter section of the doc expects FileUtil to call this method to write the file. So I am okay with this approach, as it seems the backend components are on the same page.

I think we can leverage the Customer/Account classes more here, though. In a future revision, I would expect the "practiceFile" portion of the file name to be provided though a parameter, along with a Customer object containing the object to write. I think it would be useful to have .toJson() methods on both the Customer and Account objects.

Then, this method calls Customer.toJson(), which writes the customer identifiers and calls Account.toJson() on each account object to build the final JSON representation. This whole string (or JSONObject) is returned from Customer.toJson() and written from this method.

Also, is this class going to handle reading from the filesystem? We may want to stub out a readFile method that reads from "practiceFile" and returns a JSONObject/Customer/Account.

File: data/practiceFile [hunk 7]
```diff
@@ -0,0 +1 @@
+{"Vehicle":{"Car":"Blue Tacoma"}}
```
Comment (jakecarlson1) [inline data/practiceFile:1]: I think we may want to .gitignore this directory before we have large config files here. Up to you.

File: src/main/java/data/JsonConverter.java [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+      out.close();
+    } catch(IOException e) {
+      return null;
+    } 
+
+    return file;
+  }
+
+  public boolean updateFile(Vendor vendor) {
+    createFile();
```
Comment (jakecarlson1) [inline src/main/java/data/JsonConverter.java:52]: We may just want to pass Vendor down now, so the wiring is in place. The filename can be inferred from .getVendorID() on this object.

File: src/test/java/javatests/JsonConverterTest.java [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+    } catch(FileNotFoundException e) {
+      input = null;
+    }
+
+    while(input.hasNextLine()) {
+      fileContent += input.nextLine();
+    }
+
+    // Test that the file isn't empty and has the right content.
+    assertFalse("createFile() didn't add any content to the file when is should've",
```
Comment (jakecarlson1) [inline src/test/java/javatests/JsonConverterTest.java:54]: It might be good to split this test method up into testCreateFileMethod_success, testCreateFileMethod_fails. It is a little hard to parse what is actually being tested here.

File: src/main/java/data/JsonConverter.java (hunk 6 above)
Comment (cxfloeder) [reply src/main/java/data/JsonConverter.java:28]: The idea was that the FileUtil class calls the updateFile(Vendor) method in the JsonConverter class. 

In the future, a filePath will be created using the BasePath(data/) and the vendorID. Adding a .toJson() to the Vendor and Account classes is a good idea, which I'll make in the future. 

Currently, I don't plan on having a readfile in the JsonConverter. The plan was to just write files to the filesystem.

File: data/practiceFile (hunk 7 above)
Comment (cxfloeder) [reply data/practiceFile:1]: Actually, this decision is mostly up to you. I thought we already discussed that the general format of the configs wan't a breach in privacy or important internal information. Because we don't know exactly how all these fields are actually named, I don't think that it would be necessary to put the folder in the .gitignore. The way the fields in the mock configs are named won't be the same way they are named internally.

File: src/test/java/javatests/JsonConverterTest.java (hunk 9 above)
Comment (cxfloeder) [reply src/test/java/javatests/JsonConverterTest.java:54]: In the first test, I was testing that a file was being made and that the file contained content. In the second test, I was testing that the contents of the file was the same as the data I expected would be inputted. I could definitely split these into two separate test methods.

File: src/main/java/data/JsonConverter.java (hunk 8 above)
Comment (cxfloeder) [reply src/main/java/data/JsonConverter.java:52]: To get this straight, you just want me to pass the vendorID into createFile() to create a filename based off of vendor information. I have that planned for the future, but I can quickly add that to this PR.

File: data/practiceFile (hunk 7 above)
Comment (jakecarlson1) [reply data/practiceFile:1]: I am okay with this for now. I am not concerned with the security here, just that, in the future, PRs will be harder to parse through if they have a lot of test data in them.

File: src/main/java/data/JsonConverter.java (hunk 8 above)
Comment (jakecarlson1) [reply src/main/java/data/JsonConverter.java:52]: You will probably need to pass the whole Vendor object to createFile. I am okay with this going in when you add the .toJson methods.

File: src/test/java/javatests/JsonConverterTest.java [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+      input = null;
+    }
+
+    while(input.hasNextLine()) {
+      fileContent += input.nextLine();
+    }
+
+    assertTrue("createFile() didn't add the right content to the file.", 
+      fileContent.equals(expectedFileContent));
+  }
```
Comment (ianjjohnson) [inline src/test/java/javatests/JsonConverterTest.java:83]: Consider taking the file reading piece and putting them in a separate function in this test class. That way, many tests can share this functionality (once you have more than one test)

Comment (cxfloeder) [reply src/test/java/javatests/JsonConverterTest.java:83]: That's a good idea for the next iteration. Thanks

File: src/main/java/data/JsonConverter.java (hunk 6 above)
Comment (cxfloeder) [reply src/main/java/data/JsonConverter.java]: I'll let the rest of the team know that I will add readFile functionality to the JsonConverter class on the next iteration.
